src/satellite/process.ts

```diff
@@ -633,6 +879,23 @@ export class SatelliteProcess implements Satellite {
 
     return Promise.resolve(relations)
   }
+
+  private _generateTag(timestamp: Date): string {
+    const instanceId = this._authState!.clientId
+    return generateTag(instanceId, timestamp)
+  }
+
+  async _garbageCollectOplog(transaction: Transaction) {
```

Comment:
> but we could piggyback the ack with the transaction message.
I'm not sure what you mean here. Ack to the transaction and sending transaction back are decoupled processes and the underlying system doesn't have matching between lsn and transaction.

> Second, I think it looks unsafe to rely on timestamp,
As long as timestamp is strictly monotonically increasing for the target it should be safe. If it doesn't - then conflict resolution wouldn't work properly. We can store mapping `transaction -> lsn position` on Satellite if you think that's more appropriate.

Reply:
>  Ack to the transaction and sending transaction back are decoupled processes and the underlying system doesn't have matching between lsn and transaction.

In the new approach, I think there could be no explicit ack message, because it is implied by receiving the message from the same origin. We would need to have a way to associate that with LSN, but this could be done by extending the transaction message to contain an optional ack field with the corresponding lsn position (which I don't know, but suppose is available on the server).